state-chain/pallets/cf-swapping/src/tests/fees.rs

```diff
@@ -120,80 +111,220 @@ fn test_buy_back_flip() {
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::<Test>::SwapExecuted {
+					network_fee,
+					input_amount: NORMAL_SWAP_AMOUNT,
+					..
+				}) if *network_fee == NORMAL_NETWORK_FEE * NORMAL_SWAP_AMOUNT,
+			);
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::<Test>::SwapExecuted {
+					network_fee,
+					input_amount: NORMAL_MINIMUM_NETWORK_FEE,
+					..
+				}) if *network_fee == NORMAL_MINIMUM_NETWORK_FEE,
+			);
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::<Test>::SwapExecuted {
+					network_fee,
+					input_amount: INTERNAL_SWAP_AMOUNT,
+					..
+				}) if *network_fee == INTERNAL_NETWORK_FEE * INTERNAL_SWAP_AMOUNT,
+			);
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::<Test>::SwapExecuted {
+					network_fee,
+					input_amount: INTERNAL_MINIMUM_NETWORK_FEE,
+					..
+				}) if *network_fee == INTERNAL_MINIMUM_NETWORK_FEE,
+			);
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::<Test>::SwapExecuted {
+					network_fee,
+					input_amount: NORMAL_MINIMUM_NETWORK_FEE,
+					..
+				}) if *network_fee == NORMAL_NETWORK_FEE * NORMAL_MINIMUM_NETWORK_FEE,
+			);
+
+			// Check that the network fee is actually collected
+			assert_eq!(
+				CollectedNetworkFee::<Test>::get(),
+				(NORMAL_NETWORK_FEE * NORMAL_SWAP_AMOUNT) +
+					(NORMAL_MINIMUM_NETWORK_FEE) +
+					(INTERNAL_NETWORK_FEE * INTERNAL_SWAP_AMOUNT) +
+					(INTERNAL_MINIMUM_NETWORK_FEE) +
+					(NORMAL_NETWORK_FEE * NORMAL_MINIMUM_NETWORK_FEE)
+			);
+		});
+}
 
-		let FeeTaken { remaining_amount, fee } = Swapping::take_network_fee(
-			chunk_amount,
-			MinFeePolicy::Enforced { swap_request_id: SWAP_REQUEST_ID },
+#[test]
+fn test_network_fee_tracking() {
+	new_test_ext().execute_with(|| {
+		const NETWORK_FEE: Permill = Permill::from_percent(10);
+		const MIN_NETWORK_FEE: AssetAmount = 150;
+		const CHUNK_AMOUNT: AssetAmount = 1000;
+		let normal_fee_amount = NETWORK_FEE * CHUNK_AMOUNT;
+		assert!(
+			normal_fee_amount < MIN_NETWORK_FEE,
+			"Minimum network fee must be larger than the network fee of a chunk"
 		);
 
-		// Sanity checks
-		assert_eq!(collected_fees_before + fee, CollectedNetworkFee::<Test>::get());
-		assert_eq!(remaining_amount + fee, chunk_amount);
+		// Setup a fresh tracker
+		let mut fee_tracker = NetworkFeeTracker::new(MIN_NETWORK_FEE, NETWORK_FEE);
 
+		// Take fees from each chunk in order and make sure it gives the expected result
+		// First chunk gets the minimum network fee taken from it
+		assert_eq!(
+			fee_tracker.take_fee(CHUNK_AMOUNT),
+			FeeTaken { remaining_amount: CHUNK_AMOUNT - MIN_NETWORK_FEE, fee: MIN_NETWORK_FEE }
+		);
+		// Second chunk gets partial network fee taken from it
+		assert_eq!(
+			fee_tracker.take_fee(CHUNK_AMOUNT),
+			FeeTaken {
+				remaining_amount: CHUNK_AMOUNT - (MIN_NETWORK_FEE - normal_fee_amount),
+				fee: MIN_NETWORK_FEE - normal_fee_amount
```

Comment:
This second fee should be the amount that would normally be charged from 2 chunks minus whatever you charged on the first chunk: `2 * normal_fee_amount - MIN_NETWORK_FEE`. The test currently only happens to work due to the parameters you chose, and would fail if you set `MIN_NETWORK_FEE` to 160, for example.